tools/Azure.Mcp.Tools.Search/src/Services/SearchService.cs

```diff
@@ -140,6 +143,85 @@ public async Task<List<JsonElement>> QueryIndex(
         }
     }
 
+    public async Task<List<KnowledgeSourceInfo>> ListKnowledgeSources(
+        string serviceName,
+        RetryPolicyOptions? retryPolicy = null)
+    {
+        ValidateRequiredParameters(serviceName);
+
+        var sources = new List<KnowledgeSourceInfo>();
+
+        try
+        {
+            var searchClient = await GetSearchIndexClient(serviceName, retryPolicy);
+            await foreach (var source in searchClient.GetKnowledgeSourcesAsync())
+            {
+                sources.Add(new KnowledgeSourceInfo(source.Name, source.GetType().Name, source.Description));
+            }
+            return sources;
+        }
+        catch (Exception ex)
+        {
+            throw new Exception($"Error retrieving Search knowledge sources: {ex.Message}", ex);
+        }
+    }
+
+    public async Task<List<KnowledgeBaseInfo>> ListKnowledgeBases(
+        string serviceName,
+        RetryPolicyOptions? retryPolicy = null)
+    {
+        ValidateRequiredParameters(serviceName);
+
+        var bases = new List<KnowledgeBaseInfo>();
+
+        try
+        {
+            var searchClient = await GetSearchIndexClient(serviceName, retryPolicy);
+            await foreach (var knowledgeBase in searchClient.GetKnowledgeAgentsAsync())
+            {
+                bases.Add(new KnowledgeBaseInfo(knowledgeBase.Name, knowledgeBase.Description, [.. knowledgeBase.KnowledgeSources.Select(ks => ks.Name)]));
+            }
+            return bases;
+        }
+        catch (Exception ex)
+        {
+            throw new Exception($"Error retrieving Search knowledge bases: {ex.Message}", ex);
+        }
+    }
+
+    public async Task<string> RetrieveFromKnowledgeBase(
+        string serviceName,
+        string baseName,
+        string? query,
+        IEnumerable<(string role, string message)>? messages,
+        RetryPolicyOptions? retryPolicy = null)
+    {
+        ValidateRequiredParameters(serviceName, baseName);
+
+        try
+        {
+            var searchClient = await GetSearchIndexClient(serviceName, retryPolicy);
+
+            var clientOptions = AddDefaultPolicies(new SearchClientOptions());
+            ConfigureRetryPolicy(clientOptions, retryPolicy);
+
+            var knowledgeBaseClient = new KnowledgeAgentRetrievalClient(searchClient.Endpoint, baseName, await GetCredential(), clientOptions);
+
+            var request = new KnowledgeAgentRetrievalRequest(
+                messages != null ?
+                    messages.Select(m => new KnowledgeAgentMessage([new KnowledgeAgentMessageTextContent(m.message)]) { Role = m.role }) :
+                    [new KnowledgeAgentMessage([new KnowledgeAgentMessageTextContent(query)]) { Role = "user" }]);
+
+            var response = await knowledgeBaseClient.RetrieveAsync(request);
+
+            return response.GetRawResponse().Content.ToString();
```

Comment:
Is this actually what we want to return? I ask as there could be activity records included in the raw response, and is that something useful in an agentic workflow?